feat: Handle email validation and CGU during OAuth workflow

taorepoara opened the pull request:
<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.
  
  Before submitting a Pull Request, please ensure you've done the following:
  - 👷‍♀️ Create small PRs. In most cases, this will be possible.
  - ✅ Provide tests for your changes.
  - 📜 Use Conventional Commit for your PR name (see https://www.conventionalcommits.org/en/v1.0.0/).
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation and include any relevant screenshots.
-->

<!-- 
Link the related issue
-->
Closes #

## Description of the changes
Manage email validation and CGU acceptation during the OAuth workflow.

## Checklist
- [ ] I didn't over-scope my PR
- [ ] My PR title matches the [commit convention](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] I did not include breaking changes
- [ ] I made my own code-review before requesting one


### I included unit tests that cover my changes
  - [ ] 👍 yes
  - [ ] 🙅 no, because they aren't needed
  - [ ] 🙋 no, because I need help


### I added/updated the documentation about my changes

- [ ] 📜 README.md
- [ ] 📕 docs/*.md
- [ ] 📓 docs.lenra.io
- [ ] 🙅 no documentation needed


## Technical highlight/advice
<!-- 
    Is there any complex point you want to highlight ?
    Do you need advice on specific point of your code ?
    Tell us here.
-->

taorepoara commented:
@lenra-io/devs I have random CSRF token errors. It might be due to the [`clear_session` in the `new` function](https://github.com/lenra-io/server/blob/93c6044f6c2fe6122204531e28cfac263276790a/apps/identity_web/lib/identity_web/controllers/user_auth_controller.ex#L138).

```elixir
[debug] ** (Plug.CSRFProtection.InvalidCSRFTokenError) invalid CSRF (Cross Site Request Forgery) token, please make sure that:

  * The session cookie is being sent and session is loaded
  * The request include a valid '_csrf_token' param or 'x-csrf-token' header
    (plug 1.14.2) lib/plug/csrf_protection.ex:316: Plug.CSRFProtection.call/2
    (identity_web 0.1.0) IdentityWeb.Router.browser/2
    (identity_web 0.1.0) lib/identity_web/router.ex:1: IdentityWeb.Router.__pipe_through0__/1
    (phoenix 1.6.16) lib/phoenix/router.ex:346: Phoenix.Router.__call__/2
    (identity_web 0.1.0) lib/identity_web/endpoint.ex:1: IdentityWeb.Endpoint.plug_builder_call/2
    (identity_web 0.1.0) lib/plug/debugger.ex:136: IdentityWeb.Endpoint."call (overridable 3)"/2
    (identity_web 0.1.0) lib/identity_web/endpoint.ex:1: IdentityWeb.Endpoint.call/2
    (phoenix 1.6.16) lib/phoenix/endpoint/cowboy2_handler.ex:54: Phoenix.Endpoint.Cowboy2Handler.init/4
    (cowboy 2.9.0) /home/taorepoara/lenra/server/deps/cowboy/src/cowboy_handler.erl:37: :cowboy_handler.execute/2
    (cowboy 2.9.0) /home/taorepoara/lenra/server/deps/cowboy/src/cowboy_stream_h.erl:306: :cowboy_stream_h.execute/3
    (cowboy 2.9.0) /home/taorepoara/lenra/server/deps/cowboy/src/cowboy_stream_h.erl:295: :cowboy_stream_h.request_process/3
    (stdlib 3.17.2.4) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
```

taorepoara commented:
I did not add unit tests because I don't  how to test those pages

Nesqwik reviewed (changes requested):
In addition to the comments in the code...

bugs : 
- Try to fail the login twice : "Not corresponding form data."
- Try to login, arrive to the validation code page, go back to oauth debugger, login again : CSRF token error
- Horizontal scroll when the page is too small vertically
- I don't have the "warning" icon before the error. Icon does not load. (maybe need a setup)

Preferable : 
- Keep the "challenge" (login/consent/other) in the url to keep the url consistency
- Allow to copy/paste the validation code from the first input without focusing
- The "underline" effect on "I forgot my password" seems a bit too big
- Remove the login after password reset ? It seems odd and doesn't allow to "keep login"
- The error icon is red but the text is not. Seems odd too but really personnal preference.

taorepoara commented on apps/identity_web/lib/identity_web/controllers/user_auth_controller.ex:
```diff
@@ -14,6 +18,107 @@ defmodule IdentityWeb.UserAuthController do
+      # check e-mail verification
+      user.role == :unverified_user ->
+        # send verification email
+        Accounts.resend_registration_code(user)
```
Removed

taorepoara commented on apps/identity_web/lib/identity_web/controllers/user_auth_controller.ex:
```diff
@@ -43,69 +149,200 @@ defmodule IdentityWeb.UserAuthController do
+  def cancel_login(conn, _params) do
+    login_challenge = get_session(conn, :login_challenge)
+
+    {:ok, accept_response} = HydraApi.reject_login(login_challenge, "User logged out")
```
Do you have some examples ?

taorepoara commented on apps/identity_web/lib/identity_web/controllers/user_auth_controller.ex:
```diff
@@ -43,69 +149,200 @@ defmodule IdentityWeb.UserAuthController do
 
       _error ->
         # In order to prevent user enumeration attacks, don't disclose whether the email is registered.
         render(conn, "new.html",
```
I don't think that we can keep the error messages and the changeset if we redirect. What do you think ?

Do you have a solution for that ?

jonas-martinez commented on apps/identity_web/lib/identity_web/router.ex:
```diff
@@ -18,10 +18,18 @@ defmodule IdentityWeb.Router do
     pipe_through [:browser]
 
-    get "/users/log_in", UserAuthController, :new
+    get "/users/auth", UserAuthController, :new
```
Why did you change the route? Does this have any technical reason?

taorepoara replied:
This route is common to login and register.

jonas-martinez commented on hydra/hydra.yml:
```diff
@@ -19,9 +19,9 @@ serve:
-      allowed_origins:
-        - https://oauthdebugger.com
-        - http://localhost:10000
+      # allowed_origins:
```
Uncomment or remove these lines

jonas-martinez commented on hydra/hydra.yml:
```diff
@@ -37,17 +37,17 @@ urls:
-      - public
-    pairwise:
-      salt: youReallyNeedToChangeThis
+# oidc:
```
Uncomment or remove these lines

jonas-martinez commented:
Other than that, it works very well :+1:

Nesqwik reviewed (changes requested):
- The URL in the login page still change after one fail in login or register from /users/auth?action=register to /users/register for example.
- Failing twice in a row in the login page makes the OAuth flow fail with a "User logged out" error
-

taorepoara commented:
> * The URL in the login page still change after one fail in login or register from /users/auth?action=register to /users/register for example.

This is linked to my previous comment: https://github.com/lenra-io/server/pull/394#discussion_r1267870165

I'll do it another way.

taorepoara commented:
I've fixed the issues of [the last comment](https://github.com/lenra-io/server/pull/394#pullrequestreview-1543059107) and added the password rule list.
